File: Smok.py
"""
Modeling the three smokers problem using threads and semaphores.
"""

import threading
import time
import random

# Semaphores for component pairs
TOBACCO_PAPER: threading.Semaphore = threading.Semaphore(0)
TOBACCO_FILTER: threading.Semaphore = threading.Semaphore(0)
PAPER_FILTER: threading.Semaphore = threading.Semaphore(0)
BARTENDER_SEM: threading.Semaphore = threading.Semaphore(1)


def bartender_process() -> None:
    """The flow of the bartender placing the ingredients on the table."""
    combinations: list[tuple[threading.Semaphore, str, str]] = [
        (TOBACCO_PAPER, "Табак", "Бумагу"),
        (TOBACCO_FILTER, "Табак", "Фильтр"),
        (PAPER_FILTER, "Бумагу", "Фильтр")
    ]
    while True:
        BARTENDER_SEM.acquire()
        comb_sem, item_1, item_2 = random.choice(combinations)
        print(f"\033[94mБармен положил: {item_1} и {item_2}\033[0m\n")
        comb_sem.release()


def smoker_process(name: str, has_item: str,
                   needed_sem: threading.Semaphore, needed_items: tuple[str]) -> None:
    """The flow of a smoker picking up components and smoking.

    Params:
        name (str): Smoker's name
        has_item (str): A component that the smoker already has
        needed_sem (Semaphore): A semaphore for the desired pair of components
        needed_items (tuple): Essential ingredients for smoking
    """
    while True:
        needed_sem.acquire()
        print(f"\033[92m{name} (имеет {has_item}) "
              f"взял {needed_items[0]} и {needed_items[1]}\033[0m\n")
        BARTENDER_SEM.release()

        # The smoking process
        print(f"\033[93m{name} начинает курить...\033[0m\n")
        smoke_time = random.randint(2, 4)
        time.sleep(smoke_time)

        print(f"\033[93m{name} закончил курить ({smoke_time} сек.)\033[0m\n")


if __name__ == "__main__":
    threads: list[threading.Thread] = [
        threading.Thread(target=bartender_process),
        threading.Thread(
            target=smoker_process,
            args=("Курильщик Т", "Табак", PAPER_FILTER, ("Бумагу", "Фильтр")),
            daemon=True
        ),
        threading.Thread(
            target=smoker_process,
            args=("Курильщик Б", "Бумагу", TOBACCO_FILTER, ("Табак", "Фильтр")),
            daemon=True
        ),
        threading.Thread(
            target=smoker_process,
            args=("Курильщик Ф", "Фильтр", TOBACCO_PAPER, ("Табак", "Бумагу")),
            daemon=True
        )
    ]

    for thread in threads:
        thread.start()

    for thread in threads:
        thread.join()

    try:
        while True:
            time.sleep(3600)
    except KeyboardInterrupt:
        print("\nПрограмма остановлена")

File: Smok.java
import java.util.Random;
import java.util.concurrent.Semaphore;

enum Component {
    TOBACCO, PAPER, FILTER
}

public class Smok {

    // Семафоры для пар компонентов
    private static final Semaphore tobaccoPaper = new Semaphore(0);
    private static final Semaphore tobaccoFilter = new Semaphore(0);
    private static final Semaphore paperFilter = new Semaphore(0);
    private static final Semaphore bartenderSem = new Semaphore(1);

    private static final Random random = new Random();

    public static void main(String[] args) {
        // Поток бармена
        new Thread(() -> {
            while (true) {
                try {
                    bartenderSem.acquire();

                    // Случайный выбор комбинации компонентов
                    int combination = random.nextInt(3);
                    switch (combination) {
                        case 0:
                            System.out.println("\u001B[94mБармен положил: Табак и Бумагу\u001B[0m\n");
                            tobaccoPaper.release();
                            break;
                        case 1:
                            System.out.println("\u001B[94mБармен положил: Табак и Фильтр\u001B[0m\n");
                            tobaccoFilter.release();
                            break;
                        case 2:
                            System.out.println("\u001B[94mБармен положил: Бумагу и Фильтр\u001B[0m\n");
                            paperFilter.release();
                            break;
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                }
            }
        }).start();

        // Потоки курильщиков
        new Thread(createSmoker("Курильщик Т", "Табак", paperFilter, "Бумагу", "Фильтр")).start();
        new Thread(createSmoker("Курильщик Б", "Бумагу", tobaccoFilter, "Табак", "Фильтр")).start();
        new Thread(createSmoker("Курильщик Ф", "Фильтр", tobaccoPaper, "Табак", "Бумагу")).start();

        // Бесконечное ожидание (прерывается Ctrl+C)
        try {
            Thread.sleep(Long.MAX_VALUE);
        } catch (InterruptedException e) {
            System.out.println("\nПрограмма остановлена");
        }
    }

    private static Runnable createSmoker(String name,
                                         String hasItem,
                                         Semaphore neededSem,
                                         String item1,
                                         String item2) {
        return () -> {
            while (true) {
                try {
                    neededSem.acquire();
                    System.out.printf("\u001B[92m%s (имеет %s) взял %s и %s\u001B[0m\n\n",
                            name, hasItem, item1, item2);
                    bartenderSem.release();

                    // Процесс "курения"
                    System.out.printf("\u001B[93m%s начинает курить...\u001B[0m\n\n", name);
                    int smokeTime = 2000 + random.nextInt(2000); // 2-4 секунды
                    Thread.sleep(smokeTime);

                    System.out.printf("\u001B[93m%s закончил курить (%d сек.)\u001B[0m\n\n",
                            name, smokeTime / 1000);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                }
            }
        };
    }
}
